text/3845-repr-ordered-fields.md

```diff
@@ -0,0 +1,424 @@
+If we fix our layout computation to match that of MSVC, we no longer need this special case in the ABI logic.
+
+## RFC #3718
+
+This also plays a role in [#3718](https://github.com/rust-lang/rfcs/pull/3718), where `repr(C, packed(N))` wants allow fields which are `align(M)` (while making the `repr(C, ...)` struct less packed). This is a footgun for normal uses of `repr(packed)`, so it would be better to relegate this strictly to the FFI use-case. However, since `repr(C)` plays two roles, this is difficult.
+
+By splitting `repr(ordered_fields)`  off of `repr(C)`, we can allow `repr(C, packed(N))` to contain over-aligned fields (while making the struct less packed), and (continuing to) disallow `repr(ordered_fields, packed(N))` from containing aligned fields. Thus keeping the Rust-only case free of warts, without compromising on FFI use-cases[<sup>1</sup>](#ordered_fields_align).
+
+## MSVC bug
+
+Splitting `repr(C)` also allows making progress on a workaround for the MSVC bug [rust-lang/rust/112480](https://github.com/rust-lang/rust/issues/112480). 
+
+The issue here is that MSVC is inconsistent about the alignment of `u64`/`i64` (and possibly `f64`). In MSVC, the alignment of `u64`/`i64` is reported to be 8 bytes by `alignof` and is correctly aligned in structs. However, when placed on the stack, MSVC doesn't ensure that they are aligned to 8-bytes, and may instead only align them to 4 bytes.
+
+Any proper work around will require reducing the alignment of `u64`/`i64` to 4 bytes, and adjusting what `repr(C)` to treat `u64`/`i64`'s alignment as 8 bytes. This way, if you have references/pointers to `u64`/`i64` (for example, as out pointers), then the Rust side will not break when the C side passes a 4-byte aligned pointer (but not 8-byte aligned). This could happen if the C side put the integer on the stack, or was manually allocated at some 4-byte alignment.
+
+For AIX, the issue is that `f64` is treated as aligned to 4-bytes if it is not the first field in a struct. i.e.
+```C
+struct Foo {
+	char a;
+	double b;
+}
+```
+Field `b` would be laid out at offset 4, which is under-aligned (since `f64` has alignment 8 in Rust). Again, any proper workaround will require reducing the alignment of `f64`, and adjusting `repr(C)`.
+
+## AIX layout bug
+
+For more details, see this discussion on [irlo](https://internals.rust-lang.org/t/repr-c-aix-struct-alignment/21594/3).
+
+In AIX, the following struct `Floats` has the following field offsets: `[0, 64, 96]` (in bits)
+
+```C
+struct Floats {
+    double a;
+    char b;
+    double c;
+};
+```
+
+This is because
+> In aggregates, the first member of this data type is aligned according to its natural alignment value; subsequent members of the aggregate are aligned on 4-byte boundaries.
+> - [IBM Documentation](https://www.ibm.com/docs/en/xl-c-and-cpp-aix/16.1?topic=data-using-alignment-modes) (Table 1, Note 1)
+
+Even though on AIX `alignof(double)` is 8, it is still laid out an a 4-byte boundary. 
```

Comment:
I’m not actually sure whether this is true. `__alignof__(double)` is 8 on IBM’s compiler, but `__alignof__` is an IBM compiler extension. I don’t know what the C-standard `alignof(double)` is on Clang for this target.

Reply:
Nice catch, yes I meant `__alignof__`